getScoreOfFeatureRegions<-function(list="~/jdk/rna-seq/matrix/stem.bed",
                                   MACC="~/jdk/MNase-seq1/MNase-seq1/Ara_stem_macc_U.bedGraph",
                                   upstream=1000,downstream=0,highvalue = 5,mod = 0){
  #'@description 
  #'@param list: The bed file containing the number of reads generated by features - counts
  #'@param macc: The bed file generated by the MACC package, with a score for each bin
  #'@param upstream： The distance that the promoter is located upstream of the starting site
  #'@param dowmstream: The distance at which the promoter is located downstream of the start site
  #'@param highvalue: Criteria for dividing high- and low-expression genes. It means that when the logarithmic value of TPM is greater than this parameter, the corresponding gene is considered to be highly expressed
  #'@param mod: 
  #'@example 
  a   <- getRegionsOfFeatures(upstream = upstream,downstream = downstream,txdb=TxDb.Athaliana.BioMart.plantsmart28)
  b   <- getDifferExpGeneInfo(reads_count = list,highvalue = highvalue,mod = mod)
  macc <- read.csv(file = MACC,header = T,sep = " ",col.names = c("chr","start","end","macc") )
  
  chrn_group<-split(macc,macc$chr)#根据第一列不同的值将所有行进行分组，存为列表
  high1<- a[a$gene_name %in% b[["high"]][["Geneid"]],] 
  mod1 <- a[a$gene_name %in% b[["mod"]][["Geneid"]],] 
  low1 <- a[a$gene_name %in% b[["low"]][["Geneid"]],] 
  
  high1_exp<- split(high1,high1$chrname)
  mod1_exp <- split(mod1,mod1$chrname)
  low1_exp <- split(low1,low1$chrname)

  list1<-list(high1_exp,mod1_exp,low1_exp)
  names(list1)<-c("high1_exp","mod1_exp","low1_exp")
  list2<-list()
  list3<-list()
  list4<-list()
  for (i2 in 1:length(list1)) {#for high mod low  
    
    for (i3 in 1:length(list1[[i2]])) { #for per chr  
      
      for (i4 in c(1,10,12,14)) {
        left  <-  findInterval(list1[[i2]][[i3]][[i4]],chrn_group[[i3]][[3]])+1
        righ  <-  findInterval(list1[[i2]][[i3]][[i4+1]],chrn_group[[i3]][[3]])      #寻找目标在有序向量中的位置并取该位置左侧的第一个数的位置
        right_per  <-  (as.numeric(list1[[i2]][[i3]][[i4+1]])-chrn_group[[i3]][[3]][righ])/300   #实际长度左右两侧的长度占bin的百分比
        left_per   <-  (chrn_group[[i3]][[3]][left]-as.numeric(list1[[i2]][[i3]][[i4]]))/300
        macc_value<-c()
        for (i in 1:length(list1[[i2]][[i3]][,1])) {
          
          dis  <-  righ[i]-left[i]
          if (dis<0) {
            macc_value[i] <- ((as.numeric(list1[[i2]][[i3]][[i4+1]][i])-as.numeric(list1[[i2]][[i3]][[i4]][i]))/300)*chrn_group[[i3]][[4]][righ[i]]

          }else if (dis==0) {
            r<-((as.numeric(list1[[i2]][[i3]][[i4+1]][righ[i]])-chrn_group[[i3]][[3]][righ[i]])/300)*chrn_group[[i3]][[4]][righ[i+1]]
            l<-((chrn_group[[i3]][[3]][left[i]]-as.numeric(list1[[i2]][[i3]][[i4]][left[i]]))/300)*chrn_group[[i3]][[4]][left[i]]
            macc_value[i] <- r+l

            
          }else{ #dis>0   
            mid  <-  sum(chrn_group[[i3]][[4]][c(left[i]+seq(1,dis,1))])
            macc_value[i] <- chrn_group[[i3]][[4]][left[i]]*left_per[i]+chrn_group[[i3]][[4]][righ[i]+1]*right_per[i]+mid
            if (left_per[i]>1) {
              print(i)
              print(i4)
              print(i3)
              print(i2)
              print(macc_value[i])
              print(left_per[i])
              stop("left_per[i]>1,dis>0")}
          }
          
        }
        list2[[i4]]<-macc_value
        #colnames(list1[i2][i3][i4])
      }
      list3[[i3]]<-list2
    }
    list4[[i2]]<-list3
  }
  for (ii in 1:3) {
    for (iii in 1:5) {
      list4[[ii]][[iii]][c(2,3,4,5,6,7,8,9,11,13)]<-NULL
      names(list4[[ii]][[iii]])<-c("promoter","5'","gene body","3'")
    }
    names(list4[[ii]])<-c("chr1","chr2","chr3","chr4","chr5")
  }
  names(list4)<-c("high_exp","mod_exp","low_exp")
   
  promoter<-c()
  five    <-c()
  three   <-c()
  genebody<-c()
  list6   <-c()
  list7   <-c()
  for (i in 1:length(list4)) {
    for (i1 in 1:length(list4[[i]])) {
      promoter<- c(promoter,list4[[i]][[i1]][[1]])
      five    <- c(five,    list4[[i]][[i1]][[2]])
      three   <- c(three,   list4[[i]][[i1]][[3]])
      genebody<- c(genebody,list4[[i]][[i1]][[4]])
    }
    list6[[1]] <-promoter
    list6[[2]] <-five
    list6[[3]] <-three
    list6[[4]] <-genebody
    names(list6)<-c("promoter","5'","genebody","3'")
    list7[[i]]<-list6
  }
  names(list7)<-c("high_exp","mod_exp","low_exp")  
  
  return(list7)
  
}
